File: src/c++/Calibration.cpp
#include <fstream>
#include "SeekUtil.h"
#include <libgen.h>
#include <vector>
#include <string>
#include <sstream>
#include <string.h>

#include <iostream>
#include <fstream>

extern char pallete[1001][3];
//extern char bmphdr[54];

//Calibration::Calibration () {
//    try {
//        
//    } catch (char const *err) {
//        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
//        throw;
//    }
//}

bool Calibration::execute (uint16_t *dat) {
    try {
        memcpy(raw_dat2, dat, RAW_FRAME_SIZE);
        raw_dat = dat;
        
        if (NULL == dat) {
            throw "invalid parameter";
        }
        int status = dat[10];
        if (4 == status) {
            /* gain calibration */
            //printf("gain calibration\n");
            frame4stuff();
        } else if (1 == status) {
            /* shutter calibration */
            //printf("shutter calibration\n");
            mark_badpixels();
            memcpy(arr_id1, raw_dat, sizeof(uint16_t) * RAW_FRAME_SIZE);
        } else if (3 == status) {
            /* image frame */
            //printf("image frame\n");
            
            mark_badpixels();
            frame3stuff();
            
            //for (int idx=0; idx < RAW_FRAME_SIZE; idx++) {
            //    printf("%d, ", raw_dat[idx]);
            //}
            //printf("\n");
            
            //for (int idx=0; idx < RAW_FRAME_SIZE; idx++) {
            //    
            //    printf("%d, ", arr_id3[idx]);
            //    if (arr_id3[idx] - arr_id1[idx]) {
            //        printf("%d:%d, ", arr_id1[idx], arr_id3[idx]);
            //    }
            //}
            //printf("\n");
            
            return true;
        }
        return false;
        
        //printf("init pallete\n");
        //pallete = { {0} };
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

void Calibration::frame4stuff(void) {
    try {
        int avg_id4 = getmode();
        
        for (int idx=0; idx < RAW_FRAME_SIZE; idx++) {
            if ( (raw_dat[idx] > 2000) && (raw_dat[idx] < 8000) ) {
                gain_cal[idx] = (float)avg_id4 / (float)raw_dat[idx];
                //printf("%f\n", gain_cal[idx]);
            } else {
                gain_cal[idx]  = 1;
                bad_pixel[idx] = true;
            }
        }
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

void Calibration::frame3stuff(void) {
    try {
        memcpy(arr_id3, raw_dat, sizeof(uint16_t) * RAW_FRAME_SIZE);
        
        //for (int i3=0; i3 < RAW_FRAME_SIZE ;i3++) {
        //    printf("%d, ", arr_id3[i3]);
        //}
        //printf("\n");
        
        //for (int i1=0; i1 < RAW_FRAME_SIZE ;i1++) {
        //    printf("%d,", arr_id1[i1]);
        //}
        //printf("\n");
        
        //for (int ig=0; ig < RAW_FRAME_SIZE ;ig++) {
        //    printf("%f,", gain_cal[ig]);
        //}
        //printf("\n");
        
        for (int idx=0; idx < RAW_FRAME_SIZE; idx++) {
            if (arr_id3[idx] > 2000) {
                arr_id3[idx] = (uint16_t) ((arr_id3[idx] - arr_id1[idx]) * gain_cal[idx] + 7500);
                //printf("%d,", arr_id3[idx]);
            } else {
                arr_id3[idx]   = 0;
                bad_pixel[idx] = true;
            }
        }
        
        //for (int i3=0; i3 < RAW_FRAME_SIZE ;i3++) {
        //    printf("%d,", arr_id3[i3]);
        //}
        //printf("\n");
        
        //printf("\n");
        fix_badpixels();
        
        //for (int i3=0; i3 < RAW_FRAME_SIZE ;i3++) {
        //    printf("%d,", arr_id3[i3]);
        //}
        //printf("\n");
        
        remove_noise();
        get_histogram();
        fill_imgbuff();
        
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

void Calibration::fix_badpixels(void) {
    try {
        int i   = 0;
        int nr  = 0;
        int val = 0;
        
        //for (int i3=0; i3 < RAW_FRAME_SIZE ;i3++) {
        //    printf("%d,", arr_id3[i3]);
        //}
        //printf("\n");
        
        for (int y=0; y < RAW_FRAME_HEIGHT; y++) {
            for (int x=0 ; x < RAW_FRAME_WIDTH ; x++) {
                if ( (0 != bad_pixel[i]) && (x < 206) ) {
                    val = 0;
                    nr  = 0;
                    
                    /* top pixel */
                    if ( (y > 0) && (false == bad_pixel[i - 208]) ) {
                        val += arr_id3[i - 208];
                        nr++;
                    }
                    
                    /* bottom pixel */
                    if ( (y < 155) && (false == bad_pixel[i + 208]) ) {
                        val += arr_id3[i + 208];
                        nr++;
                    }
                       
                    /* left pixel */
                    if ( (x > 0) && (false == bad_pixel[i - 1]) ) {
                        val += arr_id3[i - 1];
                        nr++;
                    }
                    
                    /* right pixel */
                    if ( (x < 205) && (false == bad_pixel[i + 1]) ) {
                        val += arr_id3[i + 1];
                        nr++;
                    }
                    
                    if (nr > 0) {
                        val = val/nr;
                        arr_id3[i] = val;
                    }
                }    
                i += 1;
            }
        }
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

void Calibration::remove_noise (void) {
    try {
        unsigned short arr_clr[] = { 0,0,0,0 };
        int val = 0;
        int i   = 0;
        
        //for (int i3=0; i3 < RAW_FRAME_SIZE ;i3++) {
        //    printf("%d,", arr_id3[i3]);
        //}
        //printf("\n");

        
        for (int y=0; y < RAW_FRAME_HEIGHT ;y++) {
            for (int x=0; x < RAW_FRAME_WIDTH ;x++) {
                
                if ( (x > 0) && (x < 206) && (y > 0) && (y < 155) ) {
                    arr_clr[0] = arr_id3[i - 208]; // top
                    arr_clr[1] = arr_id3[i + 208]; // bottom
                    arr_clr[2] = arr_id3[i - 1];   // left
                    arr_clr[3] = arr_id3[i + 1];   // right
                    
                    val = (arr_clr[0] + arr_clr[1] + arr_clr[2] + arr_clr[3] - highest(arr_clr) - lowest(arr_clr))/2;
                    if ( (abs(val - arr_id3[i]) > 100) && (val != 0) ) {
                        arr_id3[i] = val;
                    }
                    i++;
                }
            }
        }
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}


unsigned short Calibration::highest(unsigned short *num) {
    try {
        unsigned short highest = 0;
        
        for (int i=0; i < 4; i++) {
            if (num[i] > highest) {
                highest = num[i];
            }
        }
        return highest;
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

unsigned short Calibration::lowest (unsigned short *num) {
    try {
        unsigned short lowest = 30000;
        
        for (int i=0; i < 4; i++) {
            if (num[i] < lowest) {
                lowest = num[i];
            }
        }
        return lowest;
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

void Calibration::get_histogram (void) {
    try {
        //max_temp = 0;
        unsigned short arr_mode[2100] = { 0 };
        int top_pos = 0;
        
        //for (int i3=0; i3 < RAW_FRAME_SIZE ;i3++) {
        //    printf("%d,", arr_id3[i3]);
        //}
        //printf("\n");
        
        for (int idx=0; idx < RAW_FRAME_SIZE ;idx++) {
            if ( (arr_id3[idx] > 1000) && (arr_id3[idx]/10 != 0) && (false == bad_pixel[idx]) ) {
                arr_mode[arr_id3[idx]/10] += 1;
            }
        }

        //for (int arr_idx=0; arr_idx < 2100 ;arr_idx++) {
        //    printf("%d, ", arr_mode[arr_idx]);
        //}
        //printf("\n");

        /* get max of arr mode  */
        unsigned short gmode_peakcnt = 0;
        for (int am_idx=0; am_idx < 2100 ;am_idx++) {
            if (gmode_peakcnt < arr_mode[am_idx]) {
                gmode_peakcnt = arr_mode[am_idx];
            }
        }

        //printf("gmode_peakcnt:%d\n", gmode_peakcnt);
        
        /* get index of max */
        for (int am_idx2=0; am_idx2 < 2100 ;am_idx2++) {
            if (gmode_peakcnt == arr_mode[am_idx2]) {
                top_pos = am_idx2;
                break;
            }
        }

        int gmode_peakidx = top_pos * 10;
        //printf("peadidx:%d\n", gmode_peakidx);
        
        min_temp = gmode_peakidx;
        max_temp = gmode_peakidx;
        
        /* find left border */
        for (int i3=0; i3 < top_pos ;i3++) {
            if (arr_mode[i3] > arr_mode[top_pos] * 0.01) {
                min_temp = (float) i3 * 10;
                //printf("i3:%d\n", i3);
                //printf("i3*10:%d\n", i3 * 10);
                //printf("set min:%f\n", min_temp);
                break;
            }
        }
        
        /* find right border */
        for (int i4=2099; i4 > 0 ;i4--) {
            if (arr_mode[i4] > arr_mode[top_pos] * 0.01) {
                max_temp = (float) i4 * 10;
                break;
            }
        }
        
        
        //min_temp = (float) (min_temp - 5950) / 40;
        //max_temp = (float) (max_temp - 5950) / 40;
        
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

void Calibration::fill_imgbuff (void) {
    try {
        int v         = 0;
        int loc       = 0;
        float iscaler = (float) (max_temp - min_temp) / (float) 1000;
        
        //printf("iscaler: %f\n", iscaler);
        //printf("max:%f, min:%f\n", max_temp, min_temp);
        
        
        for (int i=0; i < RAW_FRAME_SIZE; i++) {
            v = arr_id3[i];
            if (v < min_temp) {
                v = min_temp;
            }
            if (v > max_temp) {
                v = max_temp;
            }
            v = v - min_temp;
            loc = (int) v / iscaler;
            
            
            bmpdat[i*3]   = pallete[loc][2];
            bmpdat[i*3+1] = pallete[loc][1];
            bmpdat[i*3+2] = pallete[loc][0];
        }
        printf("\n");
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

int Calibration::getmode(void) {
    try {
        unsigned short arr_mode[320] = { 0 };
        
        for (int idx=0; idx < RAW_FRAME_SIZE ; idx++) {
            if ( (raw_dat[idx] > 1000) && ((raw_dat[idx] / 100) != 0)) {
                arr_mode[raw_dat[idx] / 100] += 1;
            }
        }
        /* get max */
        unsigned short max = 0;
        for (int aidx=0; aidx < 320; aidx++) {
            if (max < arr_mode[aidx]) {
                max = arr_mode[aidx];
            }
        }
        /* get max index */
        for (int aidx2=0; aidx2 < 320; aidx2++) {
            if (max == arr_mode[aidx2]) {
                return aidx2 * 100;
            }
        }
        throw "could not find max index";
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}

void Calibration::mark_badpixels(void) {
    try {
	//printf("mark_badpixels\n");
        for (int idx=0; idx < RAW_FRAME_SIZE ; idx++) {
            if ((raw_dat[idx] < 2000) || (raw_dat[idx] > 22000)) {
                bad_pixel[idx] = true;
            }
        }
    } catch (char const *err) {
        cout << "[error]" << err << ": " << __FILE__ << " -> " << __LINE__ << endl;
        throw;
    }
}
/* end of file */

File: src/py/Calibration.py
import os
import csv
import array
import numpy
import sys

class Calibration:
    bmpdat  = array.array('H', [0] * 97344)
    min_val = None
    max_val = None
    
    __tf = None
    __badPixelArr = array.array('b', [0] * 32448)
    __gainCalArr  = array.array('d', [0] * 32448)
    __arrID1      = None
    __arrID3      = None
    __firstAfterCal   = False
    __lastUsableFrame = None
    __maxTempRaw      = 0
    __gMode           = 0
    __gModePeakCnt    = 0
    __gModePeakIx     = 0
    __pallete         = []

    def __init__(self):
        with open(os.path.dirname(os.path.abspath(__file__)) + '/pallete.csv') as f:
            reader = csv.reader(f)
            idx = 0
            for row in reader:
                if 0 == idx:
                    idx += 1
                    continue
                self.__pallete.append([int(row[0]),int(row[1]),int(row[2])])
            

    def execute(self,tf):
        self.__tf = tf
        ret       = False

        if 4 == tf.status:
            # gain calibration
            self.__frame4Stuff()
        elif 1 == tf.status:
            # shutter calibration
            self.__markBadPixels()
            self.__arrID1 = tf.raw_u16
            self.__firstAfterCal = True
        elif 3 == tf.status:
            # image frame
            self.__markBadPixels()
            self.__frame3Stuff()
            self.__lastUsableFrame = self.__tf
            return True
            
        return ret

    def __frame3Stuff(self):
        self.__arrID3 = self.__tf.raw_u16

        for i in range(32448):
            if self.__arrID3[i] > 2000:
                self.__arrID3[i] = int((self.__arrID3[i] - self.__arrID1[i]) * self.__gainCalArr[i] + 7500)
            else:
                self.__arrID3[i] = 0
                self.__badPixelArr[i] = True

        self.__fixBadPixels()
        self.__removeNoise()
        self.__getHistogram()
        self.__fillImgBuffer()

    def __frame4Stuff(self):
        arr_id4     = self.__tf.raw_u16
        avg_id4     = self.__getMode()
        
        for i in range(32448):
            if (arr_id4[i] > 2000) and (arr_id4[i] < 8000):
                self.__gainCalArr[i] = float(avg_id4) / float(arr_id4[i])
            else:
                self.__gainCalArr[i]  = 1
                self.__badPixelArr[i] = True
        

    def __getMode(self):
        arrMode = array.array('H', [0] * 320)
        topPos  = 0
        arr     = self.__tf.raw_u16

        for elm in arr:
            if (elm > 1000) and ((elm / 100) != 0):
                arrMode[(elm) / 100] += 1
        
        topPos = arrMode.index(numpy.amax(arrMode))
        return topPos * 100

    def __markBadPixels(self):
        raw = self.__tf.raw_u16
        for i in range(32448):
            if (raw[i] < 2000) or (raw[i] > 22000):
                self.__badPixelArr[i] = True
    
    
    def __fixBadPixels(self):
        i   = 0
        nr  = 0
        val = 0

        for y in range(156):
            for x in range(208):
                if self.__badPixelArr[i] and (x < 206):
                    val = 0
                    nr  = 0
                    
                    # top pixel
                    if (y > 0) and (False == self.__badPixelArr[i - 208]):
                        val += self.__arrID3[i - 208]
                        nr += 1

                    # bottom pixel
                    if (y < 155) and (False == self.__badPixelArr[i + 208]):
                        val += self.__arrID3[i + 208]
                        nr += 1

                    # left pixel
                    if (x > 0) and (False == self.__badPixelArr[i - 1]):
                        val += self.__arrID3[i - 1]
                        nr += 1

                    # right pixel
                    if (x < 205) and (False == self.__badPixelArr[i + 1]):
                        val += self.__arrID3[i + 1]
                        nr += 1

                    if nr > 0:
                        val = val/nr
                        self.__arrID3[i] = val

                i += 1
    
    def __removeNoise(self):
        arrColor = array.array('H', [0,0,0,0])
        val      = 0
        i        = 0

        for y in range(156):
            for x in range(208):
                if (x > 0) and (x < 206) and (y > 0) and (y < 155):
                    arrColor[0] = self.__arrID3[i - 208] # top
                    arrColor[1] = self.__arrID3[i + 208] # bottom
                    arrColor[2] = self.__arrID3[i - 1]   # left
                    arrColor[3] = self.__arrID3[i + 1]   # right

                    val = (arrColor[0] + arrColor[1] + arrColor[2] + arrColor[3] - self.__Highest(arrColor) - self.__Lowest(arrColor))/2
                    
                    if (abs(val - self.__arrID3[i]) > 100) and (val != 0):
                        self.__arrID3[i] = val

                    i += 1


    def __Highest(self,num):
        highest = 0

        for i in range(4):
            if num[i] > highest:
                highest = num[i]
        return highest

    def __Lowest(self,num):
        lowest = 30000

        for i in range(4):
            if num[i] < lowest:
                lowest = num[i]
    
        return lowest


    def __getHistogram(self):
        self.__maxTempRaw = numpy.amax(self.__arrID3)

        arrMode = array.array('H', [0] * 2100)
        topPos  = 0

        for i in range(32448):
            if (self.__arrID3[i] > 1000) and (self.__arrID3[i] / 10 != 0) and (False == self.__badPixelArr[i]):
                arrMode[self.__arrID3[i] / 10] += 1

        topPos = arrMode.index(numpy.amax(arrMode))

        self.__gMode        = arrMode
        self.__gModePeakCnt = numpy.amax(arrMode)
        self.__gModePeakIx  = topPos * 10
        
        # lower it to 100px
        for i2 in range(2100):

            self.__gMode[i2] = int(float(arrMode[i2]) / float(self.__gModePeakCnt) * 100)
        
        self.min_val = self.__gModePeakIx
        self.max_val = self.__gModePeakIx
        
        # find left border
        for i3 in range(topPos):
            if arrMode[i3] > arrMode[topPos] * 0.01:
                self.min_val = i3 * 10
                break
        
        # find right border
        i4 = 2099
        for _i4 in range(2099):
            if arrMode[i4] > arrMode[topPos] * 0.01:
                self.max_val = i4 * 10
                break
            i4 -= 1


    def __fillImgBuffer(self):
        v   = 0
        loc = 0
        iScaler = float((self.max_val - self.min_val)) / float(1000)

        for i in range(32448):
            v = self.__arrID3[i]
            if v < self.min_val:
                v = self.min_val

            if v > self.max_val:
                v = self.max_val
            
            v   = v - self.min_val
            loc = int(v / iScaler)
            
            self.bmpdat[i*3]   = self.__pallete[loc][0]
            self.bmpdat[i*3+1] = self.__pallete[loc][1]
            self.bmpdat[i*3+2] = self.__pallete[loc][2]
            
# end of file
